python/feathub/processors/flink/table_builder/join_utils.py

```diff
@@ -142,27 +146,42 @@ def __hash__(self) -> int:
 
 
 def join_table_on_key(
-    left: NativeFlinkTable, right: NativeFlinkTable, key_fields: List[str]
+    t_env: StreamTableEnvironment,
+    left: NativeFlinkTable,
+    right: NativeFlinkTable,
+    key_fields: List[str],
 ) -> NativeFlinkTable:
     """
     Join the left and right table on the given key fields.
     """
 
-    right_aliased = _rename_fields(right, key_fields)
-    predicate = _get_join_predicate(left, right_aliased, key_fields)
-
-    return left.join(right_aliased, predicate).select(
-        *[
-            native_flink_expr.col(left_field_name)
-            for left_field_name in left.get_schema().get_field_names()
-        ],
-        *[
-            native_flink_expr.col(right_field_name)
-            for right_field_name in right.get_schema().get_field_names()
-            if right_field_name not in key_fields
-        ],
+    predicate = _get_join_predicate(key_fields)
+    select_statement = ", ".join(
```

Comment:
Due to the nullability behavior gap between Table and SQL API, we need to avoid trigging casting operation through Table API in order to avoid exceptions thrown by Feathub programs. While `Table#cast` would trigger casting operation, other APIs, including joins, would also trigger implicit casing operations (casting key fields to not null). Thus this PR also migrates all join operations from Table API to SQL API.

Reply:
Would it be useful to add a comment like this:

```
Use SQL instead of Table API to get around the issue in FLINK-32464
```